engine/jobmaster/dm/ddl_coordinator.go

```diff
@@ -0,0 +1,843 @@
+			return col, nil
+		}
+	}
+	return "", nil
+}
+
+func (g *shardGroup) gcDroppedColumns(ctx context.Context) error {
+	state, err := g.droppedColumnsStore.Get(ctx)
+	if err != nil {
+		if errors.Cause(err) == metadata.ErrStateNotFound {
+			return nil
+		}
+		return err
+	}
+
+	droppedColumns := state.(*metadata.DroppedColumns)
+	cacheStmts := make(map[metadata.SourceTable]string)
+OutLoop:
+	for col := range droppedColumns.Cols {
+		for _, tbStmt := range g.normalTables {
+			if tbStmt == "" {
+				continue
+			}
+			if cols := getColumnNames(tbStmt); slices.Contains(cols, col) {
+				continue OutLoop
+			}
+		}
+
+		for sourceTable := range g.normalTables {
+			tbStmt, ok := cacheStmts[sourceTable]
+			if !ok {
+				tbStmt, err = g.tableAgent.FetchTableStmt(ctx, g.id, g.cfg, sourceTable)
+				if err != nil {
+					if strings.HasPrefix(err.Error(), "table info not found") {
+						continue
+					}
+					return err
+				}
+			}
+
+			if cols := getColumnNames(tbStmt); slices.Contains(cols, col) {
+				continue OutLoop
+			}
+		}
+		if err := g.droppedColumnsStore.DelDroppedColumn(ctx, col); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// isResoved means all tables in the group are resolved.
+// 1. no conflict ddls waiting
+// 2. all dropped column has done
+// 3. all shard tables stmts are same.
+func (g *shardGroup) isResolved(ctx context.Context) bool {
+	if len(g.conflictTables) != 0 {
+		return false
+	}
+	if _, err := g.droppedColumnsStore.Get(ctx); err == nil || errors.Cause(err) != metadata.ErrStateNotFound {
+		return false
+	}
+
+	var previousTable string
+	for _, tbStmt := range g.normalTables {
+		if tbStmt == "" {
+			continue
+		}
+		if previousTable == "" {
+			previousTable = tbStmt
+			continue
+		}
+		lhs := genCmpTable(previousTable)
+		rhs := genCmpTable(tbStmt)
+		if cmp, err := lhs.Compare(rhs); err != nil || cmp != 0 {
+			return false
+		}
+		previousTable = tbStmt
+	}
+	return true
+}
+
+func (g *shardGroup) clear(ctx context.Context) error {
+	return g.droppedColumnsStore.Delete(ctx)
+}
+
+func (g *shardGroup) showTables() map[metadata.SourceTable]ShardTable {
+	g.mu.RLock()
+	defer g.mu.RUnlock()
+	tables := make(map[metadata.SourceTable]ShardTable, 0)
+	for sourceTable, stmt := range g.normalTables {
+		tables[sourceTable] = ShardTable{
+			Current: stmt,
+			Next:    g.conflictTables[sourceTable],
+		}
+	}
+	// show dropped columns if needed.
+	return tables
+}
+
+func genCmpTable(createStmt string) schemacmp.Table {
+	p, _ := conn.GetParserFromSQLModeStr(mysql.DefaultSQLMode)
+	stmtNode, _ := p.ParseOneStmt(createStmt, "", "")
+	ti, _ := ddl.BuildTableInfoFromAST(stmtNode.(*ast.CreateTableStmt))
+	ti.State = model.StatePublic
+	return schemacmp.Encode(ti)
+}
+
+// getColumnNames and return columns' names for create table stmt.
+func getColumnNames(createStmt string) []string {
+	p, _ := conn.GetParserFromSQLModeStr(mysql.DefaultSQLMode)
+	stmtNode, _ := p.ParseOneStmt(createStmt, "", "")
```

Comment:
not sure when upstream has unsupported table structure it will cause jobmaster panic

Reply:
the createStmt was get from downstream checkpoint or from worker message, so no error here